go.mod

```diff
@@ -1,7 +1,7 @@
 module github.com/containers/skopeo
 
 // Minimum required golang version
-go 1.20 // *****  ATTENTION  WARNING  CAUTION  DANGER  ******
+go 1.21 // *****  ATTENTION  WARNING  CAUTION  DANGER  ******
```

Comment:
My understanding is that along with this, we also need `toolchain 1.21` as well?  Please double-check, and also please remove the big warning comment since it no longer applies.

Reply:
As I understand the situation (could easily be wrong), it's a complex interaction between Renovate proposed package update PRs and our CI systems.

Without the `toolchain` directive, and without any go version restrictions to Renovate, on Non-Fedora platforms (Debian, Mac, Windows), go will "auto-update" to the in-use  version of go for the package update in question (as proposed by Renovate).  Meaning for Renovate PRs, builds in CI might not be uniform across all platforms.  In other words, Windows might use golang 5.4.3 for updated package FOO 0.2 while Fedora uses golang 1.21 for FOO 0.2   I'm assuming this is a bad/unexpected thing.

OTOH, setting a `toolchain` forces the maximum go version, and both go and Renovate will honor that for non-Fedora platforms.  Meaning builds across all platforms will be exactly the same (for all shared packages).

Note: The only thing making Fedora special in these matters is they set some magic envar by default which disables the auto-update behavior.

@Luap99 please double-check my understanding, when you have a chance.

Reply:
The way I read the paragraph linked above, `toolchain` is the _minimum_ Go version; it triggers unwanted upgrades (but not downgrades).

---

… and IIRC the effect _with Renovate included_ is that at the times where Renovate adds a `toolchain` directive, it adds one that matches Renovate’s default toolchain.

… and _that_ behavior we don’t want, you’re right.

I’ll add the `toolchain` directives tomorrow. (The renovate hard-coded version will, I think, orthogonally, need bumping to 1.21.)

Reply:
yes I think we should have the toolchain there so the normal go commands can be used without adding the toolchain later, especially so renovate can be used properly.

I do however think that we need to make sure the toolchain is only ever set to the same value as the go line so we do not force update users unnecessary.

Reply:
and yes toolchain is the minimum version, not maximum.

Reply:
I have added `toolchain 1.21.0`, and updated the comment a bit.

I’m _very_ open to suggestions about the comment.

Reply:
I think the comment just needs to say:

``Warning: Ensure the `go` and `toolchain` versions match exactly to prevent unwanted auto-updates`

Reply:
Done.